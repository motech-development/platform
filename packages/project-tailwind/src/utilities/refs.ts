import { LegacyRef, MutableRefObject, useMemo, useRef } from 'react';

/**
 * Check if ref is a mutable ref object
 *
 * @param ref - Reference object or function
 *
 * @returns If ref is a mutable ref object
 */
function isMutableRefObject<T>(
  ref: MutableRefObject<T> | LegacyRef<T>,
): ref is MutableRefObject<T | null> {
  return ref != null;
}

/** useMergeRefs arguments */
type TUseMergeRefsInput<T> = Array<MutableRefObject<T> | LegacyRef<T>>;

/**
 * Combines multiple refs into a single memoized callback instance
 *
 * @param refs - Reference object or function
 *
 * @returns Combined ref object
 */
export default function useMergeRefs<T>(...refs: TUseMergeRefsInput<T>) {
  const targetRef = useRef<T>(null);

  return useMemo(() => {
    if (refs.every((ref) => ref == null)) {
      return null;
    }

    refs.forEach((ref) => {
      if (typeof ref === 'function') {
        ref(targetRef.current);
      } else if (isMutableRefObject(ref)) {
        // eslint-disable-next-line no-param-reassign
        ref.current = targetRef.current;
      }
    });

    return targetRef;
  }, [refs]);
}
