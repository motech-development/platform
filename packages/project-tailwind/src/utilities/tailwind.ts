import clsx, { ClassValue } from 'clsx';

/** Alignment options */
export enum Alignment {
  /** Centre */
  CENTRE = 'centre',
  /** Left */
  LEFT = 'left',
  /** Right */
  RIGHT = 'right',
}

/** Alignment type */
export type TAlignment = `${Alignment}`;

/** Sizing options */
export enum Sizing {
  /** Large */
  LG = 'lg',
  /** Medium */
  MD = 'md',
  /** None */
  NONE = 'none',
  /** Small */
  SM = 'sm',
}

/** Sizing type */
export type TSizing = `${Sizing}`;

/** Theming options */
export enum Themes {
  /** Danger */
  DANGER = 'danger',
  /** Primary */
  PRIMARY = 'primary',
  /** Secondary */
  SECONDARY = 'secondary',
  /** Success */
  SUCCESS = 'success',
  /** Warning */
  WARNING = 'warning',
}

/** Theme type */
export type TTheme = `${Themes}`;

/** Class names options */
type TCreateStylesClassNames = {
  /** Universal classes to apply */
  classNames?: ClassValue[];
  /** Sizing specific classes */
  sizing?: {
    [name in Sizing]?: ClassValue[];
  };
  /** Theme specific classes */
  theme?: {
    [name in TTheme]?: ClassValue[];
  };
};

/**
 * Tailwind hook
 *
 * @param theme - The theme to apply
 * @param sizing - Sizing to apply
 *
 * @returns Tailwind helper utilities
 */
export function useTailwind(theme: TTheme, sizing: TSizing) {
  /**
   * Creates css styles to apply
   *
   * @param classNames - Classes to be applied
   *
   * @returns Class names
   */
  function createStyles(classNames: TCreateStylesClassNames) {
    let finalStyles: ClassValue[] = [];

    if (classNames.classNames) {
      finalStyles = [...classNames.classNames];
    }

    if (sizing && classNames.sizing) {
      const styles = classNames.sizing[sizing];

      if (!styles) {
        throw new Error('No classes defined for this sizing.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    if (theme && classNames.theme) {
      const styles = classNames.theme[theme];

      if (!styles) {
        throw new Error('No classes defined for this theme.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    return clsx(finalStyles);
  }

  /**
   * Creates text alignment classes
   *
   * @param alignment - Text alignment
   *
   * @returns Class names
   */
  function createTextAlignmentStyles(alignment: TAlignment) {
    switch (alignment) {
      case Alignment.CENTRE:
        return 'text-center';
      case Alignment.RIGHT:
        return 'text-right';
      case Alignment.LEFT:
      default:
        return 'text-left';
    }
  }

  return {
    createStyles,
    createTextAlignmentStyles,
  };
}
