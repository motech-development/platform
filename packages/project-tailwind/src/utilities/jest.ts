/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';
import { ReactElement } from 'react';
import { Sizing, Themes } from './tailwind';

export const themes = [
  {
    theme: Themes.DANGER,
  },
  {
    theme: Themes.PRIMARY,
  },
  {
    theme: Themes.SECONDARY,
  },
  {
    theme: Themes.SUCCESS,
  },
  {
    theme: Themes.WARNING,
  },
];

export const sizing = [
  {
    size: Sizing.LG,
  },
  {
    size: Sizing.MD,
  },
  {
    size: Sizing.NONE,
  },
  {
    size: Sizing.SM,
  },
];

type TSetup = RenderResult & {
  user: UserEvent;
};

/**
 * Combines Testing Library render function with user events setup
 *
 * @param ui - React element to test
 *
 * @returns Rendered component and test user
 */
export function setup(ui: ReactElement): TSetup {
  return {
    user: userEvent.setup({
      delay: null,
    }),
    ...render(ui),
  };
}
