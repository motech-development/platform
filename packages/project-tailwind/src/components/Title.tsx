import { Themes, TTheme, useTailwind } from '../utilities/tailwind';
import { Typography } from './Typography';

/** Title component props */
export interface ITitleProps {
  /** Component theme */
  theme?: TTheme;

  /** Subtitle text */
  subtitle?: string;

  /** Title text */
  title: string;
}

/**
 * Display screen title and subtitle
 *
 * @param props - Component props
 *
 * @returns Title component
 */
export function Title({
  theme = Themes.PRIMARY,
  subtitle,
  title,
}: ITitleProps) {
  const { createStyles } = useTailwind(theme);

  const titleStyles = createStyles({
    classNames: ['decoration-clone text-gray-100 px-2 py-1'],
    theme: {
      danger: ['bg-red-600'],
      primary: ['bg-blue-600'],
      secondary: ['bg-gray-600'],
      success: ['bg-green-600'],
      warning: ['bg-yellow-600'],
    },
  });

  const subtitleStyles = createStyles({
    classNames: ['decoration-clone bg-gray-100 text-gray-900 px-2 py-1'],
  });

  return (
    <>
      <Typography asChild margin="sm" variant="h2">
        <h2>
          <span className={titleStyles}>{title}</span>
        </h2>
      </Typography>

      {subtitle && (
        <Typography margin="sm" variant="lead">
          <span className={subtitleStyles}>{subtitle}</span>
        </Typography>
      )}
    </>
  );
}
