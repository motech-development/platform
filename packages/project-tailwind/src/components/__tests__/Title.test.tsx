import { setup, themes } from '../../utilities/jest';
import { Title } from '../Title';

describe('Title', () => {
  describe.each(themes)('when theme is "$theme"', ({ theme }) => {
    it('should render the correct output', () => {
      const { asFragment } = setup(
        <Title theme={theme} title="Hello, world" />,
      );

      expect(asFragment()).toMatchSnapshot();
    });

    it('should render the correct output when a subtitle is set', () => {
      const { asFragment } = setup(
        <Title
          theme={theme}
          title="Hello, world"
          subtitle="It's nice to see you"
        />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
