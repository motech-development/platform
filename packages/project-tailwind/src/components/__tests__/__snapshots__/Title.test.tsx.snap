// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title when theme is "danger" should should render the correct output 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-red-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "danger" should should render the correct output when a subtitle is set 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-red-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
  <div>
    <p
      class="font-light text-xl text-left mb-2"
    >
      <span
        class="decoration-clone bg-gray-100 text-gray-900 px-2 py-1"
      >
        It's nice to see you
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "primary" should should render the correct output 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-blue-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "primary" should should render the correct output when a subtitle is set 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-blue-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
  <div>
    <p
      class="font-light text-xl text-left mb-2"
    >
      <span
        class="decoration-clone bg-gray-100 text-gray-900 px-2 py-1"
      >
        It's nice to see you
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "secondary" should should render the correct output 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-gray-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "secondary" should should render the correct output when a subtitle is set 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-gray-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
  <div>
    <p
      class="font-light text-xl text-left mb-2"
    >
      <span
        class="decoration-clone bg-gray-100 text-gray-900 px-2 py-1"
      >
        It's nice to see you
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "success" should should render the correct output 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-green-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "success" should should render the correct output when a subtitle is set 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-green-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
  <div>
    <p
      class="font-light text-xl text-left mb-2"
    >
      <span
        class="decoration-clone bg-gray-100 text-gray-900 px-2 py-1"
      >
        It's nice to see you
      </span>
    </p>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "warning" should should render the correct output 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-yellow-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
</DocumentFragment>
`;

exports[`Title when theme is "warning" should should render the correct output when a subtitle is set 1`] = `
<DocumentFragment>
  <div>
    <h2
      class="font-display text-4xl font-semibold text-left mb-2"
    >
      <span
        class="decoration-clone text-gray-100 px-2 py-1 bg-yellow-600"
      >
        Hello, world
      </span>
    </h2>
  </div>
  <div>
    <p
      class="font-light text-xl text-left mb-2"
    >
      <span
        class="decoration-clone bg-gray-100 text-gray-900 px-2 py-1"
      >
        It's nice to see you
      </span>
    </p>
  </div>
</DocumentFragment>
`;
